capnp-rpc/src/rpc.rs

```diff
@@ -1031,61 +1031,65 @@ impl<VatId> ConnectionState<VatId> {
-                                    }
-                                    return_::Canceled(_) => {
-                                        Self::send_unimplemented(&connection_state, &message)?;
-                                    }
-                                    return_::ResultsSentElsewhere(_) => {
-                                        Self::send_unimplemented(&connection_state, &message)?;
-                                    }
-                                    return_::TakeFromOtherQuestion(id) => {
-                                        if let Some(answer) =
-                                            connection_state.answers.borrow_mut().slots.get_mut(&id)
-                                        {
-                                            if let Some(res) = answer.redirected_results.take() {
-                                                let tmp = question_ref
-                                                    .upgrade()
-                                                    .expect("dangling question ref?");
-                                                tmp.borrow_mut().fulfill(res);
-                                            } else {
-                                                return Err(Error::failed("return.takeFromOtherQuestion referenced a call that \
-                                                     did not use sendResultsTo.yourself.".to_string()));
-                                            }
+                            Some(ref question_ref) => match ret.which()? {
+                                return_::Results(results) => {
+                                    let cap_table = Self::receive_caps(
+                                        &connection_state,
+                                        results?.get_cap_table()?,
+                                    )?;
+
+                                    let question_ref =
+                                        question_ref.upgrade().expect("dangling question ref?");
+                                    let response = Response::new(
+                                        connection_state.clone(),
+                                        question_ref.clone(),
+                                        message,
+                                        cap_table,
+                                    );
+                                    question_ref.borrow_mut().fulfill(Promise::ok(response));
+                                }
+                                return_::Exception(e) => {
+                                    let tmp =
+                                        question_ref.upgrade().expect("dangling question ref?");
+                                    tmp.borrow_mut().reject(remote_exception_to_error(e?));
+                                }
+                                return_::Canceled(_) => {
+                                    Self::send_unimplemented(&connection_state, message.as_ref())?;
```

Comment:
i've used 'as_ref' to dereference the box to the inner type.

an alternative would be to `impl IncomingMessage for Box<dyn IncomingMessage>`